### PR TITLE
1860 Corporate Bankruptcy

### PR DESCRIPTION
**Pull request by roseundy**
Last of the big corporate status implementations: Corporate Bankruptcy

Unlike closing a corporation, a bankrupt corporation keeps it's assets and can be re-parred with a wider range of prices than before it was bankrupt. Any outstanding stock shares are moved from players/pool back to IPO. A new priority deal is assigned and if the bankruptcy occurs in the middle of a stock round, the stock round is restarted. It also implements the concept of "flipped" tokens for bankrupted corps that block track laying and tokening, but not running trains:

Flipped_token

Common file changes:

- UI: Corporation: handle showing assets for bankrupt corps, shrink font for really long train strings (1860 is bad in this regard)
- UI: Token: render flipped tokens
- Engine::Token: new status instance var
- Engine::Corporation: new unfloat! method
- Round::Stock: game configuration to prevent share price change when stock is sold out
- Game::Base: add config variable for above

bankrupte corp card

### CODE REVIEW

File: assets/app/view/game/corporation.rb [hunk 1]
```diff
@@ -87,19 +87,22 @@ def render
...
+            gap: '2rem',
+            justifyContent: 'center',
+            backgroundColor: color_for(:bg2),
+            color: color_for(:font2),
+          },
+        }
 
-          children << h(:div, props, render_status)
+        if @corporation.trains.any? && !@corporation.floated?
+          if @corporation.trains.any? && !@corporation.floated?
```
Comment (tobymao) [inline assets/app/view/game/corporation.rb:101]: typo?

File: assets/app/view/game/token.rb [hunk 2]
```diff
@@ -17,6 +27,39 @@ def render
           },
         )
       end
+
+      def render_stroke
+        redw = 7
```
Comment (tobymao) [inline assets/app/view/game/token.rb:32]: constants up top

File: lib/engine/game/g_1860.rb [hunk 3]
```diff
@@ -252,11 +320,16 @@ def current_layer
...
 
       def float_corporation(corporation)
-        # Designer says that bankrupt corps keep insolvency flag
-        clear_bankrupt(corporation)
+        clear_bankrupt!(corporation)
         super
       end
 
+      def action_processed(_action)
+        @corporations.each do |corporation|
```
Comment (tobymao) [inline lib/engine/game/g_1860.rb:328]: you need to do this on every single action? that's gonna make the action * N

File: lib/engine/game/g_1860.rb [hunk 4]
```diff
@@ -450,6 +528,56 @@ def compute_distance_graph(corporation)
...
 
+      # needed for custom_node_walk
+      def custom_node_select(node, paths, corporation: nil)
+        on = paths.map { |p| [p, 0] }.to_h
+
+        custom_node_walk(node, on: on, corporation: corporation) do |path|
+          on[path] = 1 if on[path]
+        end
+
+        val = on.keys.select { |p| on[p] == 1 }
```
Comment (tobymao) [inline lib/engine/game/g_1860.rb:539]: you don't need the val, just do the selcet statement, you also don't need to do 0, just do nil, then you can do select { on[p] }

File: lib/engine/game/g_1860.rb [hunk 5]
```diff
@@ -450,6 +528,56 @@ def compute_distance_graph(corporation)
...
+        val
+      end
+
+      # needed for custom_blocks?
+      def custom_node_walk(node, visited: nil, on: nil, corporation: nil, visited_paths: {})
```
Comment (tobymao) [inline lib/engine/game/g_1860.rb:544]: what's this do?

File: assets/app/view/game/corporation.rb (hunk 1 above)
Comment (roseundy) [reply assets/app/view/game/corporation.rb:101]: Yep. Reorganized this section and biffed the cleanup.

File: assets/app/view/game/token.rb (hunk 2 above)
Comment (roseundy) [reply assets/app/view/game/token.rb:32]: OK.

File: lib/engine/game/g_1860.rb (hunk 3 above)
Comment (roseundy) [reply lib/engine/game/g_1860.rb:328]: This is already in the base class, and if anything, mine is simpler:

```
      def action_processed(_action)
        @corporations.dup.each do |corporation|
          close_corporation(corporation) if corporation.share_price&.type == :close
        end if stock_market.has_close_cell
      end
```

File: lib/engine/game/g_1860.rb (hunk 4 above)
Comment (roseundy) [reply lib/engine/game/g_1860.rb:539]: Yep, code left over from when I was debugging.

File: lib/engine/game/g_1860.rb (hunk 5 above)
Comment (roseundy) [reply lib/engine/game/g_1860.rb:544]: Fundamentally I need a version of Node::select that uses a custom City::blocks? method that understands about flipped tokens for bankrupt corporation.

Comment (roseundy) [reply lib/engine/game/g_1860.rb:544]: I thought about temporarily replacing flipped tokens with neutral tokens during route selection and dividend steps, but that seemed too inelegant and prone to bugs.

**Review by tobymao — APPROVED** [review]

**Pull request merged**